shlr/qnx/src/core.c

```diff
@@ -0,0 +1,789 @@
+	} else {
+		eprintf ("Connection failed (Protocol Version Query): %lld\n",
+			 EXTRACT_SIGNED_INTEGER (&g->recv.pkt.err.err, 4));
+		return -1;
+	}
+
+	return 0;
+}
+
+int qnxr_disconnect (libqnxr_t *g) {
+	if (!g) return -1;
+
+	if (g->connected) {
+		nto_send_init (g, DStMsg_disconnect, 0, SET_CHANNEL_DEBUG);
+		nto_send (g, sizeof(g->tran.pkt.disconnect), 0);
+		g->connected = 0;
+		g->inferior_ptid = null_ptid;
+
+		if (!r_socket_close (g->sock))
+			return -1;
+	}
+
+	return 0;
+}
+
+ptid_t qnxr_attach (libqnxr_t *g, pid_t pid) {
+
+	if (g->inferior_ptid.pid != pid) {
+		qnxr_disconnect (g);
+		sleep (1);
+		qnxr_connect (g, g->host, g->port);
+	}
+
+	nto_send_init (g, DStMsg_attach, 0, SET_CHANNEL_DEBUG);
+	g->tran.pkt.attach.pid = pid;
+	g->tran.pkt.attach.pid = EXTRACT_SIGNED_INTEGER (&g->tran.pkt.attach.pid, 4);
+
+	nto_send (g, sizeof(g->tran.pkt.attach), 0);
+	if (g->recv.pkt.hdr.cmd != DSrMsg_okdata) {
+		eprintf ("%s: failed to attach to %d\n", __func__, pid);
+		return null_ptid;
+	}
+
+	g->inferior_ptid = ptid_build (
+		EXTRACT_SIGNED_INTEGER (&g->recv.pkt.notify.pid, 4),
+		EXTRACT_SIGNED_INTEGER (&g->recv.pkt.notify.tid, 4));
+
+	return g->inferior_ptid;
+}
+
+ptid_t qnxr_run (libqnxr_t *g, const char *file, char **args, char **env) {
+	unsigned argc, envc;
+	char **argv, *p;
+	int errors = 0;
+
+	nto_send_init (g, DStMsg_env, DSMSG_ENV_CLEARENV, SET_CHANNEL_DEBUG);
+	nto_send (g, sizeof(DStMsg_env_t), 1);
+
+	for (envc = 0; *env; env++, envc++)
+		errors += !nto_send_env (g, *env);
+
+	if (errors) {
+		eprintf ("%s: error(s) occured while sending environment\n", __func__);
+	}
+
+	nto_send_init (g, DStMsg_env, DSMSG_ENV_CLEARARGV, SET_CHANNEL_DEBUG);
+	nto_send (g, sizeof(DStMsg_env_t), 1);
+
+	if (file != NULL) {
+		errors = !nto_send_arg (g, file);
+		if (!errors)
+			errors = !nto_send_arg (g, file);
+
+		if (errors) {
+			eprintf ("%s: failed to send executable file name\n", __func__);
+			return null_ptid;
+		}
+
+		errors = 0;
+		for (argv = args; *argv && **argv; argv++, argc++)
+			errors |= !nto_send_arg (g, *argv);
+
+		if (errors) {
+			eprintf ("%s: error(s) occured while sending args\n", __func__);
+		}
+	}
+
+	if (errors) return null_ptid;
+
+	nto_send_init (g, DStMsg_load, DSMSG_LOAD_DEBUG, SET_CHANNEL_DEBUG);
+	p = g->tran.pkt.load.cmdline;
+
+	g->tran.pkt.load.envc = 0;
+	g->tran.pkt.load.argc = 0;
+
+	strcpy (p, file);
+	p += strlen (p);
+	*p++ = '\0';
+
+	*p++ = '\0'; // stdin
+	*p++ = '\0'; // stdout
+	*p++ = '\0'; // stderr
+
+	nto_send (g, offsetof (DStMsg_load_t, cmdline) + p - g->tran.pkt.load.cmdline + 1, 1);
+
+	if (g->recv.pkt.hdr.cmd == DSrMsg_okdata) {
+		ptid_t ptid = nto_parse_notify (g);
+		eprintf ("%s: inferior pid: %d\n", __func__, ptid.pid);
+		g->inferior_ptid = ptid;
+
+		return ptid;
+	}
+
+	return null_ptid;
+}
+
+int qnxr_read_registers (libqnxr_t *g) {
+	if (!g) return -1;
+	int i = 0;
+	int len, rlen, regset;
+	int n = 0;
+	unsigned off;
```

Comment:
unsigned what. i know that it's int by default, but is better to make that explicit. use ut64